Replace Gzip with PGzip

MaxJa4 opened the pull request:
Closes #265 

Replaces `compress/gzip` with `github.com/klauspost/pgzip` (drop-in replacement, compatible) to significantly increase compression (archiving) performance (1.5-8+ times depending on setting `GZIP_PARALLELISM`).
Default is `1` to not use up all available compute capacity on production servers during the archiving process.

---

@m90: Also added the `--timeout 3` for `docker compose down` commands in the test suite.

MaxJa4 commented:
New test fails with `/bin/sh: /code/test/pgzip/run.sh: Permission denied` right at the beginning. It works locally. Do you know what is happening here @m90?

m90 commented:
Sorry for not mentioning this before, the `run.sh` file needs to be executable, i.e. you need to `chmod +x run.sh` still.

m90 commented on cmd/backup/archive.go:
```diff
@@ -63,7 +64,11 @@ func compress(paths []string, outFilePath, subPath string, algo string) error {
 	case "gz":
-		compressWriter = gzip.NewWriter(file)
+		w := pgzip.NewWriter(file)
+		if err := w.SetConcurrency(1<<16, concurrency); err != nil {
```
If I understand https://github.com/klauspost/pgzip/blob/17e8dac29df8ce00febbd08ee5d8ee922024a003/gzip.go#L21-L25 correctly, the default block size is `1<<20`. Why did you choose a different value? I maybe missing something, but this https://github.com/offen/docker-volume-backup/issues/265#issuecomment-1703898701 also sounds as if the default should be used.

MaxJa4 replied:
My bad, don't know how I came up with 16 there. It's supposed to be 20. Fixed that
Also changed the compression level to 5 like mentioned in the issue #265 since Gzip default (4) seems to correspond to PGzip level 5 in terms of compression ratio. So backups don't get bigger suddenly.

m90 commented on cmd/backup/config.go:
```diff
@@ -148,3 +149,21 @@ func (n *NaturalNumber) Decode(v string) error {
 	return int(*n)
 }
+
+type NaturalNumberZero int
```
```suggestion
type WholeNumber int
```
Maybe that's a bit less clunky?

MaxJa4 commented:
> the run.sh file needs to be executable
It was executable by default locally, but I had to run `git add --chmod=+x run.sh` to make git aware of that. Additionally GitHub Desktop sees a change after that, but when committing it says "error: no changes". Doing everything with command line and VS Code worked though.

m90 commented:
> but I had to run git add --chmod=+x run.sh to make git aware of that

Strange, usually git should pick this up all by itself. Could it be that you set `core.filemode` to `false` somehow?

Anyways, good to see it's running as expected now.

m90 commented on cmd/backup/config.go:
```diff
@@ -148,3 +149,21 @@ func (n *NaturalNumber) Decode(v string) error {
+		return fmt.Errorf("config: error converting %s to int", v)
+	}
+	if asInt < 0 {
+		return fmt.Errorf("config: expected a natural number, got %d", asInt)
```
```suggestion
		return fmt.Errorf("config: expected a whole number, got %d", asInt)
```

m90 commented on cmd/backup/script.go:
```diff
@@ -503,7 +504,12 @@ func (s *script) createArchive() error {
-	if err := createArchive(filesEligibleForBackup, backupSources, tarFile, s.c.BackupCompression.String()); err != nil {
+	concurrency := s.c.GzipParallelism.Int()
+	if concurrency == 0 {
+		concurrency = runtime.GOMAXPROCS(0)
```
Can this have possible side effects on following operations? Also, should it only apply when `gz` is the target compression format? Right now, this would also apply when using `zstd` compression.

I.e. should this happen further down the call stack in `compress` itself?

MaxJa4 replied:
Preferably, I'd like to put the `switch` case in `archive.go` for selecting and initializing the respective compression backend in its own function to unclutter the compress function there. It should only care about compression, not what backend is used there, as everything after the initialization is the same. Just to use the opportunity to clean up in there.

MaxJa4 commented:
I renamed `WholeNumber` again as its logically false. The name suggests that any whole number (including negative ones) are allowed - which is not the case. Since its exactly like `NaturalNumber` but with a zero, I just named it `NaturalNumber0` now to align it a little with the mathematical convention. I also added a doc string for both types to avoid confusion.

m90 commented:
Hmm, I got all my knowledge from [Wikipedia](https://en.wikipedia.org/wiki/Natural_number) which says whole numbers are natural numbers + 0. Then again, it's probably disputed and not really well defined.

MaxJa4 commented:
> Could it be that you set core.filemode to false somehow?

Not that I know of. I just changed to true, but now GIT says I have changes in all .sh files... so it definitely picked something up and was turned off before.

> which says whole numbers are natural numbers + 0

Seems like I found exactly the opposite side of that discussion haha. Well, I don't have a real preference here, now that the doc string is there to explain it.

m90 commented on test/util.sh:
```diff
@@ -35,6 +35,9 @@ docker() {
           command docker compose up --timeout 3 "$@";;
+        down)
+          shift
+          command docker compose down --timeout 3 "$@";;
```
I'm curious how this works. If a timeout for the containers is set on `up`, why does `down` need to know of it as well?

m90 reviewed (approved):
:bullettrain_side:
